pytorch_lightning/metrics/functional/object_detection.py

```diff
@@ -0,0 +1,97 @@
+
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import Counter
+
+import torch
+
+try:
+    from torchvision.ops import box_iou
+except ModuleNotFoundError:  # pragma: no-cover
+    box_iou = None
+
+
+def mean_average_precision(
+        preds: torch.Tensor, target: torch.Tensor, iou_threshold: float, ap_calculation: str
+) -> torch.Tensor:
+    """
+    Compute mean average precision for object detection task
+
+    Args:
+        preds: an Nx7 batch of predictions with representation
+               ``[image_idx, class_pred, class_prob, x_min, y_min, x_max, y_max]``
+        target: an Nx6 batch of targets with representation
+                ``[image_idx, class_label, x_min, y_min, x_max, y_max]``
+        iou_threshold: threshold for IoU score for determining true positive and
+                       false positive predictions.
+        ap_calculation: one of "step", "VOC2007", "VOC2012", or "COCO"
```

Comment:
could you elaborate a bit here. I assume that this different methods proposed by different papers. Can we insert a link to webpage that compare them?

Reply:
`step` is a step functional integral, which is how [average precision is calculated](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/metrics/functional/average_precision.py#L33-L50). `VOC2007` is an 11-point interpolation of the precision recall curve. `VOC2012` is a step functional integral with the interpolated precision recall curve. `COCO` is a 101-point interpolation of the precision recall curve. I'll link the sources for all of them soon.!

Reply:
Here are the sources as promised:
- [PASCAL VOC website](http://host.robots.ox.ac.uk/pascal/VOC/) outlines the changes to how average precision is computed, where `VOC2007` is a [TREC-style sampling approach](https://ccc.inaoep.mx/~villasen/bib/AN%20OVERVIEW%20OF%20EVALUATION%20METHODS%20IN%20TREC%20AD%20HOC%20IR%20AND%20TREC%20QA.pdf#page=15) and `VOC2010` is said to use all data points.
- I couldn't find a documentation that puts in words COCO's average precision evaluation, but their [evaluation code](https://github.com/cocodataset/cocoapi/blob/master/PythonAPI/pycocotools/cocoeval.py) has the 101-point interpolation.